chexus/_impl/models.py

```diff
@@ -27,13 +27,23 @@ class BucketItem(object):
             The object key of the S3 file object.
             This attribute is set with the name attribute if no key is
             provided.
+
+        headers (dict):
```

Comment:
From a quick read it seems like the docs, the name of this variable and the actual behavior don't match.

They're not really "headers" - what you pass in here is passed onto ExtraArgs directly, right? From https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html this seems to include a lot of stuff unrelated to HTTP headers, like ACL for instance.  If a caller *did* try to treat this as headers, they would probably call it incorrectly, since the HTTP header is "Content-Type" but the term used in ExtraArgs is different, "ContentType".

A caller can actually provide anything at all supported by ExtraArgs in the upload_file call and not only use this for Content-Type, if that's intended it should probably be documented.

This API is also difficult to use in a forwards-compatible way due to the all-or-nothing way of handling defaults.  What if the caller wants to let the default Content-Type apply but override some other upload args?  They can't do that right now.  What if you'd want to pass your own Content-Type but let any other (current and future) headers/ExtraArgs use their defaults?

Overall, I recommend not accepting arbitrary ExtraArgs but instead locking this down only to Content-Type.  I think part of the point of having publish logic put into chexus or a similar library is so that there's a central place for that logic, but if you allow the user of the library to pass arbitrary args during S3 upload then this goal is undermined because the callers can influence the upload in significant ways without chexus being designed for it (=> without chexus maintainers being aware of it).  For example, separate callers could start passing Expires args with different policies, which I think is worse than forcing this feature to be centrally designed into chexus.

Reply:
> if you allow the user of the library to pass arbitrary args during S3 upload then this goal is undermined because the callers can influence the upload in significant ways without chexus being designed for it (=> without chexus maintainers being aware of it). For example, separate callers could start passing Expires args with different policies, which I think is worse than forcing this feature to be centrally designed into chexus.

Based on this example, I also removed the functionality for the user to provide their own Content-Type argument when instantiating a BucketItem. Otherwise, we would either need to validate the Content-Type (S3 supports a long list of valid Content-Types), or at least verify that the dict contains a "ContentType" key (only validating the key seems hack-y to me). Given that our upload/publishing tools would only support a small subset of content types (and thus, Content-Types) I don't think it's too unreasonable (or, is it?) to assume that if we need to support a new content type, we could encode this custom Content-Types translation into our upload/publishing tools.